src/client.rs

```diff
@@ -582,7 +582,9 @@ where
 
         // Internal buffer, where we place messages until we have to flush
         // them to the backend.
-        let mut message_buffer = BytesMut::with_capacity(8196);
+        let mut client_message_buffer = BytesMut::with_capacity(8196);
```

Comment:
This idea came from your earlier comment about how in networking we have a buffer that we usually don't allocate and deallocate frequently. When writing to the client the server needs to populate the buffer that the client uses to send messages, which is what the `server_message_buffer` is. We were previously using the server's own buffer and then cloning it, which is an allocation and data copy.

Reply:
Yeah...this ends up being the same thing I think. Either we write to `server.buffer` or this new buffer, we have to move it out of the server buffer and clear it, and since we're clearing it, we need to clone whatever was in it in the first place.

I think to achieve what you want, we need to use slices in a very clever way. Basically when a new message comes in, we make `read_message` read that data directly into the clients buffer at the offset where we know there is no data and to ensure there is no empty bytes in between messages. We then pass the clients buffer around as a reference and whoever needs to parse the query / message inside it can do so without modifying the buffer. Finally we pass it to the server send function (borrow again) and write its contents directly to the server socket. Finally, we clear the buffer. No buffer copies, just one single copy when reading the clients message and one copy when writing it to the Postgres backend.

This approaches C-style programming honestly, and it's not straight forward. I think maybe the `bytes` crate has some handy methods for us to use? Worth a look, maybe.